reporting/transformer_example/src/transformer_example.py

```diff
@@ -50,82 +39,82 @@ class ElasticsearchRecord(object):
     doc = attrib()
 
 
-def _extract_hybrid_record(raw_record):
-    return HybridRecord(**raw_record)
+# Moving records through the pipeline ------------------------------------------
+def hybrid_record_to_data_dict(s3, hybrid_record):
+    """
+    takes an VHS/S3 location and returns its data as a record in ES format
 
+    Parameters
+    ----------
+    s3 : boto3.client
+    hybrid_record : HybridRecord
 
-def _get_record_from_s3(s3, object_location):
-    obj = s3.get_object(Bucket=object_location.namespace, Key=object_location.key)
-    return obj["Body"].read().decode("utf-8")
-
-
-def _build_es_record(s3, hybrid_record):
-    s3_record = _get_record_from_s3(s3, hybrid_record.location)
-    return ElasticsearchRecord(id=hybrid_record.id, doc=s3_record)
+    Returns
+    -------
+    data_dict : dict
+        object data in malleable dict format
+    """
+    object_location = hybrid_record.location
+    s3_object = s3.get_object(Bucket=object_location.namespace, Key=object_location.key)
+    record = s3_object["Body"].read().decode("utf-8")
+    record_data = json.loads(record)["data"]
+    return record_data
 
 
 def extract_records(s3, event):
+    """
+    extracts records from VHS and prepares them for transformation
+
+    Parameters
+    ----------
+    s3 : boto3.client
+    event : dict
+
+    Returns
+    -------
+    data_dicts : dict
+        a list of records in dict form, as specified by the input event
+    """
     raw_hybrid_records = [
         json.loads(record["Sns"]["Message"]) for record in event["Records"]
     ]
-    hybrid_records = [
-        _extract_hybrid_record(raw_record) for raw_record in raw_hybrid_records
+    hybrid_records = [HybridRecord(**record) for record in raw_hybrid_records]
+    data_dicts = [
+        hybrid_record_to_data_dict(s3, hybrid_record)
+        for hybrid_record in hybrid_records
     ]
-    return [_build_es_record(s3, hybrid_record) for hybrid_record in hybrid_records]
-
+    return data_dicts, hybrid_records
```

Comment:
You’re returning both of these because you want to combine them later, I think?

A dict would be a nicer data structure for this. And possibly not doing it all at once – how familiar are you with Python generators? Might be nice to go through some of those together at some point, it would clean up a bunch of this code.

(If you get a “TypeError: unhashable type: ‘HybridRecord’”, that’s fixable. I can go through it with you.)

Reply:
wrt generators, were you thinking something along these lines?

```python
def extract_records(s3, event):
    for record in event["Records"]:
        raw_hybrid_record = json.loads(record["Sns"]["Message"])
        hybrid_record = HybridRecord(**raw_hybrid_record)
        s3_object = fetch_object_from_s3(s3, hybrid_record)
        data_dict = s3_object_to_data_dict(s3_object)
        yield data_dict, hybrid_record
```

 happy to wait until friday to chat in person if you'd prefer

Reply:
Along the right lines, but I’d decompose it into a bunch of separate generators that you chain together. More like:

```python
def extract_sns_messages_from_event(event):
    for record in event["Records"]:
        yield json.loads(record["Sns"]["Message"])


def get_hybrid_objects_from_messages(s3, messages):
    for msg in messages:
        hybrid_record = HybridRecord(**raw_hybrid_record)
        yield hybrid_record, fetch_object_from_s3(s3, hybrid_record)


def get_data_dicts_from_hybrid_objects(hybrid_objects):
    for hybrid_record, s3_obj in hybrid_objects:
        data_dict = s3_object_to_data_dict(s3_object)
        yield data_dict, hybrid_record


if __name__ == "__main__":
    messages = extract_sns_messages_from_event(event)
    hybrid_objects = get_hybrid_objects_from_messages(s3, messages)
    data_dicts = get_data_dicts_from_hybrid_objects(hybrid_objects)
```

Each of the individual generators is much easier to iterate on or test, and then you have reusable generators you can glue together into a different variant of the pipeline later.

Reply:
Much neater now, I hope! Thanks for the advice